In [1]:
import pandas as pd

In [2]:
pd.set_option('display.max_columns', None)
df = pd.read_csv('/content/all_season_details.csv')
df.head()

<ipython-input-2-c286dfe52521>:2: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/all_season_details.csv')


comment_id  season  match_id match_name home_team away_team  \
0         110  2023.0   1359475   GT v CSK        GT       CSK   
1         120  2023.0   1359475   GT v CSK        GT       CSK   
2         130  2023.0   1359475   GT v CSK        GT       CSK   
3         140  2023.0   1359475   GT v CSK        GT       CSK   
4         150  2023.0   1359475   GT v CSK        GT       CSK   

  current_innings  innings_id  over  ball  runs  \
0             CSK           1     1     1     0   
1             CSK           1     1     2     1   
2             CSK           1     1     3     0   
3             CSK           1     1     4     1   
4             CSK           1     1     5     0   

                             shortText  isBoundary  isWide  isNoball  \
0     Mohammed Shami to Conway, no run       False   False     False   
1  Mohammed Shami to Conway, 1 leg bye       False   False     False   
2    Mohammed Shami to Gaikwad, no run       False   False     False   
3     Mohammed Shami to Gaikwad, 1 run       False   False     False   
4     Mohammed Shami to Conway, no run       False   False     False   

   batsman1_id    batsman1_name  batsman1_runs  batsman1_balls  bowler1_id  \
0       379140     Devon Conway              0               1      481896   
1       379140     Devon Conway              0               2      481896   
2      1060380  Ruturaj Gaikwad              0               1      481896   
3      1060380  Ruturaj Gaikwad              1               2      481896   
4       379140     Devon Conway              0               3      481896   

     bowler1_name  bowler1_overs  bowler1_maidens  bowler1_runs  bowler1_wkts  \
0  Mohammed Shami            0.1                0             0             0   
1  Mohammed Shami            0.2                0             0             0   
2  Mohammed Shami            0.3                0             0             0   
3  Mohammed Shami            0.4                0             1             0   
4  Mohammed Shami            0.5                0             1             0   

  batsman2_id    batsman2_name  batsman2_runs  batsman2_balls  bowler2_id  \
0     1060380  Ruturaj Gaikwad              0               0         NaN   
1     1060380  Ruturaj Gaikwad              0               0         NaN   
2      379140     Devon Conway              0               2         NaN   
3      379140     Devon Conway              0               2         NaN   
4     1060380  Ruturaj Gaikwad              1               2         NaN   

  bowler2_name  bowler2_overs  bowler2_maidens  bowler2_runs  bowler2_wkts  \
0          NaN            NaN              NaN           NaN           NaN   
1          NaN            NaN              NaN           NaN           NaN   
2          NaN            NaN              NaN           NaN           NaN   
3          NaN            NaN              NaN           NaN           NaN   
4          NaN            NaN              NaN           NaN           NaN   

   wicket_id wkt_batsman_name wkt_bowler_name  wkt_batsman_runs  \
0        NaN              NaN             NaN               NaN   
1        NaN              NaN             NaN               NaN   
2        NaN              NaN             NaN               NaN   
3        NaN              NaN             NaN               NaN   
4        NaN              NaN             NaN               NaN   

   wkt_batsman_balls wkt_text  isRetiredHurt  \
0                NaN      NaN          False   
1                NaN      NaN          False   
2                NaN      NaN          False   
3                NaN      NaN          False   
4                NaN      NaN          False   

                                                text  \
0  nice and full, angling and perhaps swinging in...   
1  Conway shuffles across off, and shimmies out a...   
2  good length, angling in at off and then straig...   
3  shorter and slanting into middle and leg, Gaik...   
4  Sha

In [3]:
df.isnull().sum()

comment_id                0
season                  483
match_id                  0
match_name                0
home_team                 0
away_team                 0
current_innings           0
innings_id                0
over                      0
ball                      0
runs                      0
shortText                 0
isBoundary                0
isWide                    0
isNoball                  0
batsman1_id               0
batsman1_name             0
batsman1_runs             0
batsman1_balls            0
bowler1_id                0
bowler1_name              0
bowler1_overs             0
bowler1_maidens           0
bowler1_runs              0
bowler1_wkts              0
batsman2_id               0
batsman2_name             2
batsman2_runs             0
batsman2_balls            0
bowler2_id            12841
bowler2_name          12841
bowler2_overs         12841
bowler2_maidens       12841
bowler2_runs          12841
bowler2_wkts          12841
wicket_id            230670
wkt_batsman_name     230670
wkt_bowler_name      230670
wkt_batsman_runs     230670
wkt_batsman_balls    230670
wkt_text             230670
isRetiredHurt             0
text                   1372
preText              196275
postText             222856
dtype: int64

In [4]:
df['home_team'].value_counts()

home_team
CSK      29624
SRH      29061
DC       28900
RCB      28478
MI       28333
KKR      26725
RR       22804
KXIP     21994
PWI       5478
PBKS      4817
GT        4369
LSG       3598
GL        3526
RPS       3284
Kochi     1559
Name: count, dtype: int64

Some of the teams are not required as they are not playing till now.

So training with the that teams makes nothing better but increases the complexity and time consuming

In [6]:
print(df.info)

<bound method DataFrame.info of         comment_id  season  match_id match_name home_team away_team  \
0              110  2023.0   1359475   GT v CSK        GT       CSK   
1              120  2023.0   1359475   GT v CSK        GT       CSK   
2              130  2023.0   1359475   GT v CSK        GT       CSK   
3              140  2023.0   1359475   GT v CSK        GT       CSK   
4              150  2023.0   1359475   GT v CSK        GT       CSK   
...            ...     ...       ...        ...       ...       ...   
242545      219020  2008.0    336040   CSK v RR       CSK        RR   
242546      219030  2008.0    336040   CSK v RR       CSK        RR   
242547      219050  2008.0    336040   CSK v RR       CSK        RR   
242548      219040  2008.0    336040   CSK v RR       CSK        RR   
242549      219060  2008.0    336040   CSK v RR       CSK        RR   

       current_innings  innings_id  over  ball  runs  \
0                  CSK           1     1     1     0   
1  

In [7]:
print(df.duplicated().sum())

1


In [ ]:
df['innings_id'].value_counts()

innings_id
1    125339
2    117211
Name: count, dtype: int64

This Shows that first innings and seconds innings doesn't match

The short text indicates how many runs each bowler conceded to a specific batsman on a particular ball.

In [ ]:
df['shortText'].value_counts()

shortText
Sharma to Kohli, 1 run                     94
Kumar to Gayle, no run                     76
Kumar to Smith, no run                     71
Kumar to Rahane, no run                    70
Harbhajan Singh to Raina, 1 run            66
                                           ..
Bumrah to Uthappa, 2 runs                   1
Archer to Sharma, 1 wide                    1
HH Pandya to Shakib Al Hasan, FOUR runs     1
HH Pandya to Shakib Al Hasan, OUT           1
Balaji to Sohail Tanvir, 2 runs             1
Name: count, Length: 88537, dtype: int64

In [ ]:
df.shape

(242550, 45)

In [ ]:
df['runs'].value_counts()

runs
1    100884
0     84694
4     27888
2     16064
6     11595
3       878
5       470
7        77
Name: count, dtype: int64



*   In one ball only 3 runs we can score but the dataset is showing runs upto 7
*   If we need to score the runs more than 3 we can hit boundary.
*   In our case whenever the runs are 4 and 6 we can see that the for some rows the isBoundary value is False
*   How can we get 4,5,6,7 runs if it is not boundary.



In [ ]:
df.groupby(['runs','isBoundary'])['isBoundary'].count()

runs  isBoundary
0     False          84694
1     False         100884
2     False          16064
3     False            878
4     False            487
      True           27401
5     False            351
      True             119
6     False              6
      True           11589
7     True              77
Name: isBoundary, dtype: int64

**Evaluation on the 7 runs:**

In [ ]:
temp = df.loc[df['runs'] == 7, :]
temp['isWide'].value_counts()

isWide
False    77
Name: count, dtype: int64

In [ ]:
temp['isNoball'].value_counts()

isNoball
True    77
Name: count, dtype: int64

In [ ]:
temp.shape

(77, 45)

In [ ]:
temp.head()

comment_id  season  match_id  match_name home_team away_team  \
449       213050  2023.0   1359476  PBKS v KKR      PBKS       KKR   
959       219030  2023.0   1359478    SRH v RR       SRH        RR   
1956       14020  2023.0   1359483   KKR v RCB       KKR       RCB   
2042      117060  2023.0   1359483   KKR v RCB       KKR       RCB   
3708      117050  2023.0   1359490     DC v MI        DC        MI   

     current_innings  innings_id  over  ball  runs  \
449              KKR           2    14     4     7   
959              SRH           2    20     3     7   
1956             KKR           1     5     2     7   
2042             KKR           1    18     5     7   
3708              DC           1    18     5     7   

                                            shortText  isBoundary  isWide  \
449                      Ellis to Iyer, (no ball) SIX        True   False   
959               Saini to Umran Malik, (no ball) SIX        True   False   
1956  Akash Deep to Rahmanullah Gurbaz, (no ball) SIX        True   False   
2042           Mohammed Siraj to Singh, (no ball) SIX        True   False   
3708                 Meredith to Patel, (no ball) SIX        True   False   

      isNoball  batsman1_id       batsman1_name  batsman1_runs  \
449       True       851403      Venkatesh Iyer             31   
959       True      1246528         Umran Malik             18   
1956      True       974087  Rahmanullah Gurbaz             19   
2042      True       723105         Rinku Singh             29   
3708      True       554691          Axar Patel             53   

      batsman1_balls  bowler1_id    bowler1_name  bowler1_overs  \
449               24      826915    Nathan Ellis            2.3   
959                7      700167   Navdeep Saini            1.2   
1956              15     1176959      Akash Deep            0.1   
2042              26      940973  Mohammed Siraj            2.4   
3708              22      961713  Riley Meredith            2.4   

      bowler1_maidens  bowler1_runs  bowler1_wkts batsman2_id   batsman2_name  \
449                 0            25             1      276298   Andre Russell   
959                 0            25             0     1175485     Abdul Samad   
1956                0             7             0      604527     Nitish Rana   
2042                0            32             0      475281  Shardul Thakur   
3708                0            27             1      219889    David Warner   

      batsman2_runs  batsman2_balls  bowler2_id       bowler2_name  \
449              23              13   1064812.0       Rahul Chahar   
959              24              29    277912.0        Trent Boult   
1956              0               3    308251.0       David Willey   
2042             63              25    390481.0      Harshal Patel   
3708             51              46    272477.0  Jason Behrendorff   

      bowler2_overs  bowler2_maidens  bowler2_runs  bowler2_wkts  wicket_id  \
449             2.0              0.0          12.0           1.0        NaN   
959             4.0              1.0          21.0           2.0        NaN   
1956            2.0              1.0           3.0           2.0        NaN   
2042            2.0              0.0          21.0           0.0        NaN   
3708            2.0              0.0          22.0           0.0        NaN   

     wkt_batsman_name wkt_bowler_name  wkt_batsman_runs  wkt_batsman_balls  \
449               NaN             NaN               NaN                NaN   
959               NaN             NaN               NaN                NaN   
1956              NaN             NaN               NaN                NaN   
2042              NaN             NaN               NaN                NaN   
3708              NaN             NaN               NaN                NaN   

     wkt_text  isRetiredHurt  \
449       NaN          False   
959       NaN          False   
1956      NaN          False   
2042      NaN    

In [ ]:
temp.loc[temp[['match_id','innings_id','over','ball','runs']].duplicated(),:]

comment_id  season  match_id match_name home_team away_team  \
50591      119030  2020.0   1216496   RR v CSK        RR       CSK   

      current_innings  innings_id  over  ball  runs  \
50591              RR           1    20     3     7   

                                 shortText  isBoundary  isWide  isNoball  \
50591  Ngidi to Archer, (no ball) SIX runs        True   False      True   

       batsman1_id batsman1_name  batsman1_runs  batsman1_balls  bowler1_id  \
50591       669855  Jofra Archer             19               4      542023   

      bowler1_name  bowler1_overs  bowler1_maidens  bowler1_runs  \
50591  Lungi Ngidi            3.2                0            45   

       bowler1_wkts batsman2_id batsman2_name  batsman2_runs  batsman2_balls  \
50591             1      550235    Tom Curran              9               8   

       bowler2_id bowler2_name  bowler2_overs  bowler2_maidens  bowler2_runs  \
50591    662973.0   Sam Curran            4.0              0.0          33.0   

       bowler2_wkts  wicket_id wkt_batsman_name wkt_bowler_name  \
50591           3.0        NaN              NaN             NaN   

       wkt_batsman_runs  wkt_batsman_balls wkt_text  isRetiredHurt  \
50591               NaN                NaN      NaN          False   

                                                    text preText postText  
50591  That is three in a row. Holly Molly. Short of ...     NaN      NaN

In [ ]:
temp[temp[['match_id','innings_id','over','ball','runs']].duplicated()].index

Index([50591], dtype='int64')

In [ ]:
df[df.duplicated()].index

Index([13082], dtype='int64')

**Evaluation on the 6 runs**

In [ ]:
temp6 = df.loc[(df['runs'] == 6) & (df['isBoundary'] == False), :]
temp6['isWide'].value_counts()

isWide
False    6
Name: count, dtype: int64

In [ ]:
temp6['isNoball'].value_counts()

isNoball
False    6
Name: count, dtype: int64

**Its a rare condition:**


*   6 runs are scored by running between the wickets
*   These happens because of the mistakes in fielding(overthrows,etc..)



In [ ]:
temp6.head()

comment_id  season  match_id match_name home_team away_team  \
63734        16040  2020.0   1237180   DC v SRH        DC       SRH   
107573       29070  2016.0    980909   KKR v MI       KKR        MI   
163883      210030  2013.0    598061  KKR v PWI       KKR       PWI   
167046      211030  2012.0    548308   MI v PWI        MI       PWI   
182905       27010  2012.0    548376   DC v RCB       SRH       RCB   

       current_innings  innings_id  over  ball  runs  \
63734               DC           1     7     4     6   
107573              MI           2    10     6     6   
163883             KKR           2    11     3     6   
167046              MI           2    12     3     6   
182905             RCB           2     8     1     6   

                              shortText  isBoundary  isWide  isNoball  \
63734     Rashid Khan to Dhawan, 6 runs       False   False     False   
107573  Kuldeep Yadav to Sharma, 6 runs       False   False     False   
163883   Yuvraj Singh to Pathan, 6 runs       False   False     False   
167046        Kumar to Franklin, 6 runs       False   False     False   
182905          Mishra to Kohli, 6 runs       False   False     False   

        batsman1_id   batsman1_name  batsman1_runs  batsman1_balls  \
63734         28235  Shikhar Dhawan             37              18   
107573        34102    Rohit Sharma             45              31   
163883        32498    Yusuf Pathan             31              19   
167046        37004  James Franklin             24              26   
182905       253802     Virat Kohli             14              13   

        bowler1_id       bowler1_name  bowler1_overs  bowler1_maidens  \
63734       793463        Rashid Khan            0.4                0   
107573      559235      Kuldeep Yadav            2.0                0   
163883       36084       Yuvraj Singh            0.3                0   
167046      326016  Bhuvneshwar Kumar            1.3                0   
182905       31107        Amit Mishra            0.1                0   

        bowler1_runs  bowler1_wkts batsman2_id       batsman2_name  \
63734              9             0      325012      Marcus Stoinis   
107573            20             0      625371       Hardik Pandya   
163883             9             0       47660  Ryan ten Doeschate   
167046            12             0      446507    Suryakumar Yadav   
182905             6             0       44936      AB de Villiers   

        batsman2_runs  batsman2_balls  bowler2_id       bowler2_name  \
63734              35              23     31872.0     Shahbaz Nadeem   
107573              9               9     32966.0      Piyush Chawla   
163883             23              16     26490.0        Ali Murtaza   
167046              0               1     30049.0      Murali Kartik   
182905              2               4    528967.0  Veer Pratap Singh   

        bowler2_overs  bowler2_maidens  bowler2_runs  bowler2_wkts  wicket_id  \
63734             1.0              0.0          14.0           0.0        NaN   
107573            2.0              0.0          14.0           0.0        NaN   
163883            2.0              0.0          17.0           0.0        NaN   
167046            2.0              0.0           8.0           2.0        NaN   
182905            1.0              0.0           4.0           0.0        NaN   

       wkt_batsman_name wkt_bowler_name  wkt_batsman_runs  wkt_batsman_balls  \
63734               NaN             NaN               NaN                NaN   
107573              NaN             NaN               NaN                NaN   
163883              NaN             NaN               NaN                NaN   
167046              NaN             NaN               NaN                NaN   
182905              NaN             NaN               NaN                NaN   

       wkt_text  isRetiredHurt  \
63734       NaN          False   
107573      NaN          False   
163883      NaN      

**Evaluation on the 5 runs:**

In [ ]:
temp5 = df.loc[(df['runs'] == 5) & (df['isBoundary'] == False), :]
temp5['isWide'].value_counts()

isWide
True     279
False     72
Name: count, dtype: int64

In [ ]:
temp5['isNoball'].value_counts()

isNoball
False    345
True       6
Name: count, dtype: int64

In [ ]:
temp5.shape

(351, 45)

In [ ]:
temp5.head()

comment_id  season  match_id match_name home_team away_team  \
709       218060  2023.0   1359477   LSG v DC       LSG        DC   
1197       11010  2023.0   1359480  CSK v LSG       CSK       LSG   
1350       25010  2023.0   1359480  CSK v LSG       CSK       LSG   
1446         180  2023.0   1359481    DC v GT        DC        GT   
1859       28060  2023.0   1359482  RR v PBKS        RR      PBKS   

     current_innings  innings_id  over  ball  runs  \
709               DC           2    19     6     5   
1197             CSK           1     2     1     5   
1350             LSG           2     6     1     5   
1446              DC           1     1     6     5   
1859              RR           2     9     5     5   

                              shortText  isBoundary  isWide  isNoball  \
709   Unadkat to Kuldeep Yadav, 5 wides       False    True     False   
1197     Avesh Khan to Gaikwad, 5 wides       False    True     False   
1350              Ali to Mayers, 5 runs       False   False     False   
1446  Mohammed Shami to Warner, 5 wides       False    True     False   
1859   Sikandar Raza to Samson, 5 wides       False    True     False   

      batsman1_id    batsman1_name  batsman1_runs  batsman1_balls  bowler1_id  \
709        559235    Kuldeep Yadav              6               9      390484   
1197      1060380  Ruturaj Gaikwad              3               3      694211   
1350       348056      Kyle Mayers             53              21        8917   
1446       219889     David Warner              4               5      481896   
1859       425943     Sanju Samson             34              17      299572   

        bowler1_name  bowler1_overs  bowler1_maidens  bowler1_runs  \
709   Jaydev Unadkat            2.5                0            39   
1197      Avesh Khan            0.0                0             5   
1350       Moeen Ali            0.1                0             5   
1446  Mohammed Shami            0.5                0            11   
1859   Sikandar Raza            0.4                0             9   

      bowler1_wkts batsman2_id     batsman2_name  batsman2_runs  \
709              0      554691        Axar Patel             16   
1197             0      379140      Devon Conway              2   
1350             0      422108          KL Rahul             18   
1446             0     1070168      Prithvi Shaw              0   
1859             0     1119026  Devdutt Padikkal              7   

      batsman2_balls  bowler2_id   bowler2_name  bowler2_overs  \
709               10    694211.0     Avesh Khan            4.0   
1197               3    348056.0    Kyle Mayers            1.0   
1350              12    447261.0  Deepak Chahar            3.0   
1446               0         NaN            NaN            NaN   
1859              12   1064812.0   Rahul Chahar            1.0   

      bowler2_maidens  bowler2_runs  bowler2_wkts  wicket_id wkt_batsman_name  \
709               0.0          29.0           2.0        NaN              NaN   
1197              0.0           6.0           0.0        NaN              NaN   
1350              0.0          37.0           0.0        NaN              NaN   
1446              NaN           NaN           NaN        NaN              NaN   
1859              0.0           7.0           0.0        NaN              NaN   

     wkt_bowler_name  wkt_batsman_runs  wkt_batsman_balls wkt_text  \
709              NaN               NaN                NaN      NaN   
1197             NaN               NaN                NaN      NaN   
1350             NaN               NaN                NaN      NaN   
1446             NaN               NaN                NaN      NaN   
1859             NaN               NaN                NaN      NaN   

      isRetiredHurt                                               text  \
709           False  bouncer follows Kuldeep backing away from leg ...   
1197          False  full and swinging well down leg past a fli

In [ ]:
index_of_either_wide_or_noball = temp5[((temp5['isWide'] == True) & (temp5['isNoball'] == False)) | (temp5['isWide'] == False) & (temp5['isNoball'] == True)]
len(index_of_either_wide_or_noball)

285

In [ ]:
index_of_not_wide_and_noball = temp5[(temp5['isWide'] == False) & (temp5['isNoball'] == False)]
len(index_of_not_wide_and_noball)

66

In [ ]:
temp5.shape

(351, 45)

285 rows + 66 rows = 351

**The 66 rows are also the rare condition**

In [ ]:
index_of_not_wide_and_noball.head()

comment_id  season  match_id  match_name home_team away_team  \
1350        25010  2023.0   1359480   CSK v LSG       CSK       LSG   
49137       23030  2021.0   1254060   SRH v KKR       SRH       KKR   
58315       13060  2020.0   1216517   MI v KXIP        MI      KXIP   
58661         240  2020.0   1216533    CSK v RR       CSK        RR   
65534       25010  2019.0   1175361  KKR v KXIP       KKR      KXIP   

      current_innings  innings_id  over  ball  runs  \
1350              LSG           2     6     1     5   
49137             SRH           2     4     3     5   
58315              MI           1     4     6     5   
58661              RR           2     1     4     5   
65534            KXIP           2     6     1     5   

                                shortText  isBoundary  isWide  isNoball  \
1350                Ali to Mayers, 5 runs       False   False     False   
49137   Prasidh Krishna to Pandey, 5 runs       False   False     False   
58315   Mohammed Shami to de Kock, 5 runs       False   False     False   
58661            Chahar to Stokes, 5 runs       False   False     False   
65534  Prasidh Krishna to Agarwal, 5 runs       False   False     False   

       batsman1_id    batsman1_name  batsman1_runs  batsman1_balls  \
1350        348056      Kyle Mayers             53              21   
49137       290630    Manish Pandey              5               4   
58315       379143  Quinton de Kock             18              10   
58661       311158       Ben Stokes              9               4   
65534       398438   Mayank Agarwal             20              10   

       bowler1_id     bowler1_name  bowler1_overs  bowler1_maidens  \
1350         8917        Moeen Ali            0.1                0   
49137      917159  Prasidh Krishna            1.3                0   
58315      481896   Mohammed Shami            2.0                0   
58661      447261    Deepak Chahar            0.4                0   
65534      917159  Prasidh Krishna            2.1                0   

       bowler1_runs  bowler1_wkts batsman2_id   batsman2_name  batsman2_runs  \
1350              5             0      422108        KL Rahul             18   
49137            10             1      297433  Jonny Bairstow              4   
58315            15             1      720471    Ishan Kishan              1   
58661             9             0       35582   Robin Uthappa              0   
65534            29             0      642525   Sarfaraz Khan              5   

       batsman2_balls  bowler2_id     bowler2_name  bowler2_overs  \
1350               12    447261.0    Deepak Chahar            3.0   
49137               7     56143.0  Shakib Al Hasan            1.0   
58315               2   1125976.0   Arshdeep Singh            1.0   
58661               0         NaN              NaN            NaN   
65534               3    276298.0    Andre Russell            1.0   

       bowler2_maidens  bowler2_runs  bowler2_wkts  wicket_id  \
1350               0.0          37.0           0.0        NaN   
49137              0.0           1.0           1.0        NaN   
58315              0.0           8.0           1.0        NaN   
58661              NaN           NaN           NaN        NaN   
65534              0.0           7.0           1.0        NaN   

      wkt_batsman_name wkt_bowler_name  wkt_batsman_runs  wkt_batsman_balls  \
1350               NaN             NaN               NaN                NaN   
49137              NaN             NaN               NaN                NaN   
58315              NaN             NaN               NaN                NaN   
58661              NaN             NaN               NaN                NaN   
65534              NaN             NaN               NaN                NaN   

      wkt_text  isRetiredHurt  \
1350       NaN          False   
49137      NaN          False   
58315      NaN          False   
58661      NaN          False   
65534      NaN          False